samples/sensor/sht3xd/src/main.c

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2018 Peter Bigot Consulting, LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <sensor.h>
+#include <stdio.h>
+
+#ifdef CONFIG_SHT3XD_TRIGGER
+static void trigger_handler(struct device *dev, struct sensor_trigger *trig)
+{
+	struct sensor_value temp, hum;
+	int rc;
+
+	rc = sensor_sample_fetch(dev);
+	if (rc == 0) {
+		rc = sensor_channel_get(dev, SENSOR_CHAN_AMBIENT_TEMP, &temp);
+	}
+	if (rc == 0) {
+		rc = sensor_channel_get(dev, SENSOR_CHAN_HUMIDITY, &hum);
+	}
+	if (rc != 0) {
+		printf("SHT3XD.TRIGGER: failed: %d\n", rc);
+		/* SHT3XD alert is implemented as a level interrupt
+		 * and sensor_sample_fetch() will fail when this is
+		 * re-invoked before the next sample is available.  A
+		 * real application should probably reconfigure the
+		 * threshold to detect it going below a lower value.
+		 * This one just blocks. */
+		while (true) {
+		}
+	} else {
+		printf("SHT3XD.TRIGGER: %.2f Cel ; %0.2f %%RH\n",
+		       sensor_value_to_double(&temp),
+		       sensor_value_to_double(&hum));
+	}
+}
+#endif
+
+void main(void)
+{
+	struct device *dev = device_get_binding("SHT3XD");
+	int rc;
+
+	if (dev == NULL) {
+		printf("Could not get SHT3XD device\n");
+		return;
+	}
+
+#ifdef CONFIG_SHT3XD_TRIGGER
+	struct sensor_value val;
+	struct sensor_trigger trig;
+
+	val.val1 = 21;
+	val.val2 = 0;
+
+	rc = sensor_attr_set(dev, SENSOR_CHAN_AMBIENT_TEMP,
+			     SENSOR_ATTR_UPPER_THRESH, &val);
+	if (0 == rc) {
+		trig.type = SENSOR_TRIG_THRESHOLD;
+		trig.chan = SENSOR_CHAN_AMBIENT_TEMP;
+		rc = sensor_trigger_set(dev, &trig, trigger_handler);
+	}
+
+	if (0 != rc) {
+		printf("SHT3XD.TRIGGER: failed to configure: %d\n", rc);
+		return;
+	}
+	printf("SHT3XD.TRIGGER above for %0.2f Cel\n",
+	       sensor_value_to_double(&val));
+#endif
+
+	while (true) {
+		struct sensor_value temp, hum;
+
+		rc = sensor_sample_fetch(dev);
+		if (rc == 0) {
+			rc = sensor_channel_get(dev, SENSOR_CHAN_AMBIENT_TEMP, &temp);
```

Comment:
It doesn't make sense to fetch and get the sensor data in the trigger handler, then do it again here. Can you refactor such that the trigger handler just gives a semaphore?

Reply:
It got a bit complex but I've replaced it with something a little more realistic.  All issues should be resolved; let me know if you want something more.